### PR TITLE
implement first half

### PR DESCRIPTION
**Pull request by rachelrobins**

### CODE REVIEW

File: src/ast/AstLlvmPrintVisitor.java [hunk 1]
```diff
@@ -81,6 +83,7 @@ public void visit(MainClass mainClass) {
     @Override
     public void visit(MethodDecl methodDecl) {
         // TODO EX2 - print declaration
+        builder.append(methodDecl);
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: I don't think this is the decleration.
It should be 'declare ...' in llvm format

File: src/ast/AstLlvmPrintVisitor.java [hunk 2]
```diff
@@ -124,10 +131,25 @@ public void visit(FormalArg formalArg) {
...
     public void visit(VarDecl varDecl) {
         // TODO EX2 - %varDecl.name = alloca varDecl.type().accept(this);
+
         appendWithIndent("");
-        varDecl.type().accept(this);
-        builder.append(" ");
-        builder.append(varDecl.name());
+        builder.append("%"+varDecl.name());
+        builder.append("alloca ");
+        builder.append("=");
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java:178]: alloca should be after the '='

File: src/ast/AstLlvmPrintVisitor.java [hunk 3]
```diff
@@ -165,10 +209,34 @@ public void visit(WhileStatement whileStatement) {
...
+            builder.append("\n");
+
+        }
+        else{
+            whileStatement.cond().accept(this);
+            builder.append("br i1 %_"+ getLastRegisterCount()+", label %if"+whileLabel+", label %if"+outLabel);
+        }
         indent++;
-        whileStatement.body().accept(this);
+        //whileStatement.body().accept(this);
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: why did you comment this out?

File: src/ast/AstLlvmPrintVisitor.java [hunk 4]
```diff
@@ -165,10 +209,34 @@ public void visit(WhileStatement whileStatement) {
...
         whileStatement.cond().accept(this);
-        builder.append(") {");
+        if(whileStatement.cond() instanceof TrueExpr){
+            builder.append("br label %if"+whileLabel);
+            builder.append("\n");
+        }
+
+        if(whileStatement.cond() instanceof FalseExpr){
+
+            builder.append("br label %if"+whileLabel);
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: what is the difference between true and false? shouldn't it be 'outLabel'?

File: src/ast/AstLlvmPrintVisitor.java [hunk 5]
```diff
@@ -179,25 +247,108 @@ public void visit(SysoutStatement sysoutStatement) {
...
+        if( sysoutStatement.arg() instanceof  AddExpr || sysoutStatement.arg() instanceof  SubtractExpr ||  sysoutStatement.arg() instanceof  MultExpr || sysoutStatement.arg() instanceof ArrayAccessExpr) {
+            builder.append("call i32 @puts(i32 " + "%_"+getLastRegisterCount()+")");
+            builder.append("\n");
+        }
+        if( sysoutStatement.arg() instanceof  MethodCallExpr) {
+            //Todo : access symboltable to get type
+        }
+
+        if( sysoutStatement.arg() instanceof  TrueExpr ) {
+            builder.append("call i32 @puts(i1 " + "true");
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: what is "true"?
should it be "1"?

File: src/ast/AstLlvmPrintVisitor.java [hunk 6]
```diff
@@ -179,25 +247,108 @@ public void visit(SysoutStatement sysoutStatement) {
...
+
+        //Todo : update currentlvalue
         assignStatement.rv().accept(this);
+        if(assignStatement.rv() instanceof AddExpr || assignStatement.rv() instanceof  SubtractExpr || assignStatement.rv() instanceof  MultExpr || assignStatement.rv() instanceof ArrayAccessExpr  ){
+            builder.append("store i32 %_" +getLastRegisterCount()+", i32* %"+assignStatement.lv());
+            builder.append("\n");
+        }
+        if( assignStatement.rv() instanceof  IntegerLiteralExpr) {
+            //print-int(sysoutStatement.arg());
+            builder.append("store i32 %_" +assignStatement.rv()+", i32* %"+assignStatement.lv());
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: I think it should be '.rv().num()'

File: src/ast/AstLlvmPrintVisitor.java [hunk 7]
```diff
@@ -179,25 +247,108 @@ public void visit(SysoutStatement sysoutStatement) {
...
+            builder.append("\n");
+        }
+        if( assignStatement.rv() instanceof  LtExpr) {
+            builder.append("store i1 %_" +getLastRegisterCount()+", i1 %"+assignStatement.lv());
+            builder.append("\n");
+        }
+        if( assignStatement.rv() instanceof  TrueExpr ) {
+            builder.append("store 1 " + ", i1 %"+assignStatement.lv());
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: should be 'store i1 1' - right?

File: src/ast/AstLlvmPrintVisitor.java [hunk 8]
```diff
@@ -213,7 +364,65 @@ public void visit(AssignArrayStatement assignArrayStatement) {
...
+
+        }
+        builder.append("\n");
+        builder.append("if"+andcond1+":");
+        builder.append("\n");
+        e.e2().accept(this);
+        if(e.e2() instanceof  TrueExpr){
+            regCnt++;
+            assignedval=regCnt;
+            builder.append("%_"+getLastRegisterCount() +"= 1");
```
Comment (chenmoryosef) [inline src/ast/AstLlvmPrintVisitor.java]: maybe "= i1 1" ?

**Review by chenmoryosef — CHANGES_REQUESTED** [review]
Nice! See comments :)

File: src/ast/AstLlvmPrintVisitor.java (hunk 1 above)
Comment (galiheim) [reply src/ast/AstLlvmPrintVisitor.java]: "define RETTYPE @CLASSNAME.METHODENAME(i8* %this, OPTIONAL-MORE-ARGS)"

**Comment by chenmoryosef** [comment]
@rachelrobins @galiheim please make sure what I did here is good.
ef9c8a6

**Review by chenmoryosef — APPROVED** [review]

**Pull request merged**